ilcs_parser/__init__.py

```diff
@@ -0,0 +1,173 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+import pycrfsuite
+import os
+import re
+import warnings
+from collections import OrderedDict
+import string
+
+# Labels are based on the Illinois Compiled Statutes numbering scheme:
+# https://ilga.gov/commission/lrb/lrbnew.htm#ILCS
+# SubSection and Attempted are included even though they are not technically
+# part of the spec, since it's often useful to compare these tokens
+
+LABELS = [
+    'Chapter',
+    'ActPrefix',
+    'Section',
+    'SubSection',
+    'Attempted'
+]
+
+PARENT_LABEL  = 'Citation'
+GROUP_LABEL   = 'CitationCollection'
+NULL_LABEL    = 'Null'
+MODEL_FILE    = 'learned_settings.crfsuite'
+
+
+try:
+    TAGGER = pycrfsuite.Tagger()
+    TAGGER.open(os.path.split(os.path.abspath(__file__))[0]+'/'+MODEL_FILE)
+except IOError:
+    TAGGER = None
+    warnings.warn(
+        'You must train the model (parserator train [traindata] [modulename]) '
+        'to create the %s file before you can use the parse and tag methods' % MODEL_FILE
+    )
+
+
+def parse(raw_string):
+    if not TAGGER:
+        msg = '\nMISSING MODEL FILE: %s\n' % MODEL_FILE
+        msg += (
+            'You must train the model before you can use the parse and tag '
+            'methods\nTo train the model annd create the model file, run:\n'
+            'parserator train [traindata] [modulename]'
+        )
+        raise IOError(msg)
+
+    tokens = tokenize(raw_string)
+    if not tokens:
+        return []
+
+    features = tokens2features(tokens)
+
+    tags = TAGGER.tag(features)
+
+    # Strip semantic slashes from tokens
+    tokens = [re.sub("/", "", token) for token in tokens]
+
+    return list(zip(tokens, tags))
+
+
+def tag(raw_string):
+    tagged = OrderedDict()
+    for token, label in parse(raw_string):
+        tagged.setdefault(label, []).append(token)
+
+    for token in tagged:
+        component = ' '.join(tagged[token])
+        component = component.strip(' ,;')
+        tagged[token] = component
+
+    return tagged
+
+
+def tokenize(raw_string):
+    if isinstance(raw_string, bytes):
+        try:
+            raw_string = str(raw_string, encoding='utf-8')
+        except:
+            raw_string = str(raw_string)
+
+    # Replace back slashes with forward slashes
+    raw_string = re.sub(r"\\", "/", raw_string)
+
+    # Remove any instances of the string "ILCS"
+    raw_string = re.sub(r"ilcs", " ", raw_string, flags=re.IGNORECASE)
+
+    # Remove instances of stray IDs formatted like "(tp337049)"
+    raw_string = re.sub(r"\(tp[^\)]+\)", " ", raw_string, flags=re.IGNORECASE)
+
+    re_tokens = re.compile(r"""
```

Comment:
probably too brittle, but fine place to start

Reply:
I agree, it's been a pain to work on this regex even just in this branch. I basically took the equivalent regexes from [probablepeople](https://github.com/datamade/probablepeople/blob/672075cb23a86321d35b3b407b3f2d5e2dcadfa4/probablepeople/__init__.py#L158-L165) and [usaddress](https://github.com/datamade/usaddress/blob/40fff23afaf0144f6fbcf55c112cc033a638b05b/usaddress/__init__.py#L211-L216) and adapted them to ILCS codes, but if you have thoughts about a more flexible approach I'd be interested to hear it.

Reply:
if we were reliably able to identify the separators the problem would be a lot simpler. I'd be tempted to treat the separators as semantically meaningful, like maybe "ChapterSeparator" or something like that.

then tokenize by strings of contiguous numbers and letters and their separators.

so "'625-5/11-501(a)(2)" would be tokenized to

[625, -, 5, /, 11, -, 501, (, a, ), (, 2, )]

does that make sense?

Reply:
an alternate approach, that could also work well, is to just tokenize to sequences of contiguous numbers and letters 

so "'625-5/11-501(a)(2)" would be tokenized to

[625, 5, 11, 501, a, 2]

i actually suspect that would do better.

Reply:
> if we were reliably able to identify the separators the problem would be a lot simpler. I'd be tempted to treat the separators as semantically meaningful, like maybe "ChapterSeparator" or something like that.

Yeah, that's very smart. It also gets around the issue I had where I wanted `/` to be meaningful (it's nearly always something we might call `ActSectionSeparator`) so I wound up bundling it in with whatever token followed it and then just stripping it out in the `parse()` method; it works but is semantically incorrect since e.g. the `Section` is really `11` not `/11`.

> an alternate approach, that could also work well, is to just tokenize to sequences of contiguous numbers and letters

This is basically what we're doing now, we're just preserving `/` and parentheses because they felt particularly semantically meaningful. But I would be interested in trying it with no punctuation and seeing how performance compares.